Migrate transaction handling to libwally

cdecker opened the pull request:
This is my attempt to dogfood the `libwally` transaction handling in c-lightning. Ultimately it'll remove quite a bit of code that we need to maintain and allow us to get upstream support from the wally team. This is definitely not the easiest (or shortest) PR ever, but I did my best to keep it reviewable (see below).

I tried to structure the commits in such a way that they are easy to review
and could be tested at every small change in order to guarantee
correctness. This means that this PR takes the scenic route to its goal:

 - The first few commits deal with constructing both the old style transaction
   and the wally transaction in parallel. A number of shims for adding outputs
   and inputs were added.
 - Followed by a few commits adding assertions that everything works as
   expected, i.e., the two transactions were being generated identical to each
   other (cfr. `bitcoint_tx_check` and hash equality assertions when signing). 
 - Then the binaries are moved off of using `tx->input` and `tx->output` since
   these are now managed by `libwally` internally.
 - Finally we rip out the `tx->input` and `tx->output` handling and everything
   should be back to normal.

## Open questions

I'm not sure whether leaving the shims in is a good idea yet, but I like the
idea of having `bitcoin/tx.c` be the boundary between `wally` and the rest of
the code, so maybe this verbose code is good to stay?

## Reviewing / Merging

Abandon all hope, ye who enter :door: 

I kept most fixups separate, but in squash order, so reviewing becomes less
painful. Many fixups only concern a single file and could potentially be done
in a separate commit, but I felt that'd be over the top, so I separated most
things by binary they produce.

Once we agree that the PR is good, I'll squash the commits, to keep a minimal
set that has logical boundaries.

cdecker commented:
ackbot?

niftynei commented:
i know you're planning to squash/redo how these commits appear but 

ACK https://github.com/ElementsProject/lightning/pull/2500/commits/f1366046fd3a0e22b800dd808d1be259d1a72a08

cdecker commented on hsmd/hsmd.c:
```diff
@@ -773,7 +773,8 @@ static struct io_plan *handle_sign_commitment_tx(struct io_conn *conn,
 	 * you'll crash if you assume it's there and you're wrong. */
-	tx->input[0].amount = tal_dup(tx->input, struct amount_sat, &funding);
+	tx->input_amounts[0] = tal_dup(tx->input, struct amount_sat, &funding);
+	tx->input_amounts[0] = tx->input_amounts[0];
```
Squashing will resolve this :-)

cdecker commented:
> i know you're planning to squash/redo how these commits appear but

Yeah sorry about the unsquashed commits, I was hoping it'd make things easier to review, but it meant leaving some intermediate fixes dangling. All commits with `fixup!` as  a prefix of the commit message will be squashed into the corresponding parent commit :wink:

cdecker commented:
Awesome, I'll fix and squash ASAP :-)

cdecker commented on bitcoin/test/run-tx-encode.c:
```diff
@@ -6,7 +6,13 @@
 #include <common/utils.h>
 
-const char extended_tx[] = "02000000000101b5bef485c41d0d1f58d1e8a561924ece5c476d86cff063ea10c8df06136eb31d00000000171600144aa38e396e1394fb45cbf83f48d1464fbc9f498fffffffff0140330f000000000017a9140580ba016669d3efaf09a0b2ec3954469ea2bf038702483045022100f2abf9e9cf238c66533af93f23937eae8ac01fb6f105a00ab71dbefb9637dc9502205c1ac745829b3f6889607961f5d817dfa0c8f52bdda12e837c4f7b162f6db8a701210204096eb817f ...
+const char extended_tx[] =
```
I don't quite follow, I just broke the line into 78 character lines, so it'd be nicer to read.

cdecker commented:
Rebased and squashed. Ready for the next review barrage @rustyrussell  and @niftynei :-)

 - Range-diff of the squash: https://gist.github.com/cdecker/5de66f923b5f7c90be4bbb140b161c0a (very noisy due to fixups).
 - Range-diff of the rebase on master https://gist.github.com/cdecker/9f7ebfe609e6a0507321625f12d37b64

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/pull/2500/commits/74562cabe1f988d52b07aa7120ec734fe6a3b804

rustyrussell commented:
Let's not spend too much time on (re-re-re-)review.  Applied.
